dev/TeachingTip/TeachingTip.cpp

```diff
@@ -1208,249 +1255,391 @@ void TeachingTip::StartContractToClose()
+            leftTopAvailable = false;
+        }
+        // If the tip is too tall to fit in the window when the tail is centered vertically on the target and the tip.
+        if (contentHeight / 2.0f > screenSpaceAroundTarget.Top + (targetBounds.Height / 2.0f) ||
+            contentHeight / 2.0f > screenSpaceAroundTarget.Bottom + (targetBounds.Height / 2.0f))
+        {
+            rightCenterAvailable = false;
+            leftCenterAvailable = false;
+        }
+        // If the tip is too tall to fit between the center of the target and the bottom edge of the window.
+        if (contentHeight - MinimumTipEdgeToTailCenter() > screenSpaceAroundTarget.Bottom + (targetBounds.Height / 2.0f))
+        {
+            rightBottomAvailable = false;
+            leftBottomAvailable = false;
+        }
+        // If the tip is too tall to fit between the bottom of the target and the bottom edge of the window.
+        if (tipHeight > screenSpaceAroundTarget.Bottom)
+        {
+            bottomCenterAvailable = false;
+            bottomLeftAvailable = false;
+            bottomRightAvailable = false;
+        }
+
+        // If the tip is too wide to fit between the left edge of the target and the left edge of the window.
+        if (tipWidth > screenSpaceAroundTarget.Left)
+        {
+            leftCenterAvailable = false;
+            leftTopAvailable = false;
+            leftBottomAvailable = false;
+        }
+        // If the tip is too wide to fit between the center of the target and the left edge of the window.
+        if (contentWidth - MinimumTipEdgeToTailCenter() > screenSpaceAroundTarget.Left + (targetBounds.Width / 2.0f))
+        {
+            topLeftAvailable = false;
+            bottomLeftAvailable = false;
+        }
+        // If the tip is too wide to fit in the window when the tail is centered horizontally on the target and the tip.
+        if (contentWidth / 2.0f > screenSpaceAroundTarget.Left + (targetBounds.Width / 2.0f) ||
+            contentWidth / 2.0f > screenSpaceAroundTarget.Right + (targetBounds.Width / 2.0f))
+        {
+            topCenterAvailable = false;
+            bottomCenterAvailable = false;
+        }
+        // If the tip is too wide to fit between the center of the target and the right edge of the window.
+        if (contentWidth - MinimumTipEdgeToTailCenter() > screenSpaceAroundTarget.Right + (targetBounds.Width / 2.0f))
+        {
+            topRightAvailable = false;
+            bottomRightAvailable = false;
+        }
+        // If the tip is too wide to fit between the right edge of the target and the right edge of the window.
+        if (tipWidth > screenSpaceAroundTarget.Right)
+        {
+            rightCenterAvailable = false;
+            rightTopAvailable = false;
+            rightBottomAvailable = false;
+        }
+
+        auto priorities = GetPlacementFallbackOrder(placement);
+
+        for each (winrt::TeachingTipPlacementMode mode in priorities)
+        {
```

Comment:
"for each" is an old MSVC extension. range-for is standard now:
for (auto mode in priorities)